<a href="https://colab.research.google.com/github/rhRohitHalder/ml-practice/blob/main/SONAR_Rock_vs_Mine_binary_class_logistic_regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd ## loading data
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

### DATA COLLECTION AND DATA PREPROCESSING

In [2]:
df = pd.read_csv('/content/sonar data.csv', header=None)

In [3]:
df.shape

(208, 61)

In [4]:
df.describe() ## describe ---? statistical measures of data

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
count,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,...,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000
mean,0.029164,0.038437,0.043832,0.053892,0.075202,0.104570,0.121747,0.134799,0.178003,0.208259,...,0.016069,0.013420,0.010709,0.010941,0.009290,0.008222,0.007820,0.007949,0.007941,0.006507
std,0.022991,0.032960,0.038428,0.046528,0.055552,0.059105,0.061788,0.085152,0.118387,0.134416,...,0.012008,0.009634,0.007060,0.007301,0.007088,0.005736,0.005785,0.006470,0.006181,0.005031
min,0.001500,0.000600,0.001500,0.005800,0.006700,0.010200,0.003300,0.005500,0.007500,0.011300,...,0.000000,0.000800,0.000500,0.001000,0.000600,0.000400,0.000300,0.000300,0.000100,0.000600
25%,0.013350,0.016450,0.018950,0.024375,0.038050,0.067025,0.080900,0.080425,0.097025,0.111275,...,0.008425,0.007275,0.005075,0.005375,0.004150,0.004400,0.003700,0.003600,0.003675,0.003100
50%,0.022800,0.030800,0.034300,0.044050,0.062500,0.092150,0.106950,0.112100,0.152250,0.182400,...,0.013900,0.011400,0.009550,0.009300,0.007500,0.006850,0.005950,0.005800,0.006400,0.005300
75%,0.035550,0.047950,0.057950,0.064500,0.100275,0.134125,0.154000,0.169600,0.233425,0.268700,...,0.020825,0.016725,0.014900,0.014500,0.012100,0.010575,0.010425,0.010350,0.010325,0.008525
max,0.137100,0.233900,0.305900,0.426400,0.401000,0.382300,0.372900,0.459000,0.682800,0.710600,...,0.100400,0.070900,0.039000,0.035200,0.044700,0.039400,0.035500,0.044000,0.036400,0.043900


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       208 non-null    float64
 1   1       208 non-null    float64
 2   2       208 non-null    float64
 3   3       208 non-null    float64
 4   4       208 non-null    float64
 5   5       208 non-null    float64
 6   6       208 non-null    float64
 7   7       208 non-null    float64
 8   8       208 non-null    float64
 9   9       208 non-null    float64
 10  10      208 non-null    float64
 11  11      208 non-null    float64
 12  12      208 non-null    float64
 13  13      208 non-null    float64
 14  14      208 non-null    float64
 15  15      208 non-null    float64
 16  16      208 non-null    float64
 17  17      208 non-null    float64
 18  18      208 non-null    float64
 19  19      208 non-null    float64
 20  20      208 non-null    float64
 21  21      208 non-null    float64
 22  22

In [6]:
### Class Distribution
df[60].value_counts()

,count
60,
M,111
R,97


1. There are 208 observations (samples/rows).Each row is one sonar signal.
2. Target = Column 60
3. What am I trying to predict? - R/M -> Binary Classification
4. mine - 111 , Rock - 97 , That's nearly balanced.

Don't just run df.info() and move on. Ask yourself these questions every time:

How many rows are there?


1.   How many rows are there?
2.   How many columns are there?
3.   Which column is the target?
4.   What is the target's data type?
5.   Are there missing values?
6. Are features numerical or categorical?
7. Is this a classification or regression problem?
8. Is the dataset small or large?
9. Will scaling or encoding be needed?
10. Which ML algorithms are likely to fit this dataset?


1. How many rows are there?
  There are 208 rows (observations) in this dataset.

2. How many columns are there?
There are 61 columns total (60 input features and 1 output/target variable).

3. Which column is the target?
Column 60 (the last column) is the target.

4. What is the target's data type?
The target's data type is object (it contains the text categories "R" for Rock and "M" for Mine).

5. Are there missing values?
No, there are zero missing values. Every column has 208 non-null entries, perfectly matching the total row count.

6. Are features numerical or categorical?
All 60 features are purely numerical (float64).

7. Is this a classification or regression problem?
This is a binary classification problem because the goal is to predict between two specific categories (R or M).

8. Is the dataset small or large?
The dataset is very small (only 208 rows and about 99 KB in memory).

9. Will scaling or encoding be needed?

Encoding: No feature encoding is needed since all inputs are numerical. You only need to encode the target label (R/M) into 0s and 1s.

Scaling: Yes, scaling (like StandardScaler) is highly recommended since ML algorithms like SVM, KNN, and Logistic Regression are sensitive to feature scales.

10. Which ML algorithms are likely to fit this dataset?
    Because it is a small, numerical dataset, traditional machine learning algorithms like Logistic Regression, Support Vector Machines (SVM), Random Forest, Decision Trees, K-Nearest Neighbors (KNN), and Naive Bayes are ideal. Deep learning is not suitable here due to the lack of massive data.

In [7]:
### What is the average value of each numerical column for each group?
df.groupby(60).mean()

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
60,,,,,,,,,,,,,,,,,,,,,
M,0.034989,0.045544,0.050720,0.064768,0.086715,0.111864,0.128359,0.149832,0.213492,0.251022,...,0.019352,0.016014,0.011643,0.012185,0.009923,0.008914,0.007825,0.009060,0.008695,0.006930
R,0.022498,0.030303,0.035951,0.041447,0.062028,0.096224,0.114180,0.117596,0.137392,0.159325,...,0.012311,0.010453,0.009640,0.009518,0.008567,0.007430,0.007814,0.006677,0.007078,0.006024


In [8]:
# separating based on data and label
# There are two axes:
# axis=0 → Rows (vertical)
# axis=1 → Columns (horizontal)

X = df.drop(columns=60, axis=1)
Y = df[60]

In [9]:
print(X)  # dataframe only with features
print(Y)  # data only with lable

         0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      50      51      52      53   

training and test data

In [10]:
# straitify=Y ,This ensures the class proportions are maintained in both the training and testing sets.
X_train , x_test , Y_train , y_test = train_test_split(X,Y,test_size=0.1,stratify=Y,random_state=42)

In [11]:
print(X.shape,X_train.shape,x_test.shape)

(208, 60) (187, 60) (21, 60)


In [12]:
print(Y.shape,Y_train.shape,y_test.shape)

(208,) (187,) (21,)


In [13]:
print(X_train)
print(Y_train)

         0       1       2       3       4       5       6       7       8   \
138  0.0731  0.1249  0.1665  0.1496  0.1443  0.2770  0.2555  0.1712  0.0466   
161  0.0305  0.0363  0.0214  0.0227  0.0456  0.0665  0.0939  0.0972  0.2535   
201  0.0335  0.0258  0.0398  0.0570  0.0529  0.1091  0.1709  0.1684  0.1865   
122  0.0249  0.0119  0.0277  0.0760  0.1218  0.1538  0.1192  0.1229  0.2119   
147  0.0654  0.0649  0.0737  0.1132  0.2482  0.1257  0.1797  0.0989  0.2460   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
116  0.0094  0.0333  0.0306  0.0376  0.1296  0.1795  0.1909  0.1692  0.1870   
39   0.0091  0.0213  0.0206  0.0505  0.0657  0.0795  0.0970  0.0872  0.0743   
23   0.0115  0.0150  0.0136  0.0076  0.0211  0.1058  0.1023  0.0440  0.0931   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53   

model training -> logistic regresion


In [14]:
model = LogisticRegression()

In [16]:
# training the logistic Regression model with training data
# "Learn the relationship between the input features (X_train) and the correct answers (Y_train)."
model.fit(X_train , Y_train)

LogisticRegression()

Model Evaluation

In [23]:
#accuracy on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction,Y_train )

In [24]:
print('Accuracy on training : ', training_data_accuracy)

Accuracy on training :  0.8235294117647058


In [25]:
#accuracy on testing data
X_testing_prediction = model.predict(x_test)
testing_data_accuracy = accuracy_score(X_testing_prediction,y_test )

In [27]:
print('Accuracy on test data : ', testing_data_accuracy)

Accuracy on test data :  0.8571428571428571


Making a Predictive System

In [30]:
input_data = (0.1088,0.1278,0.0926,0.1234,0.1276,0.1731,0.1948,0.4262,0.6828,0.5761,0.4733,0.2362,0.1023,0.2904,0.4713,0.4659,0.1415,0.0849,0.3257,0.9007,0.9312,0.4856,0.1346,0.1604,0.2737,0.5609,0.3654,0.6139,0.5470,0.8474,0.5638,0.5443,0.5086,0.6253,0.8497,0.8406,0.8420,0.9136,0.7713,0.4882,0.3724,0.4469,0.4586,0.4491,0.5616,0.4305,0.0945,0.0794,0.0274,0.0154,0.0140,0.0455,0.0213,0.0082,0.0124,0.0167,0.0103,0.0205,0.0178,0.0187)

# changing the input_data to a numpy array
input_data_as_numpy_array = np.asarray(input_data)

# reshape the np array as we are predicting for one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

prediction = model.predict(input_data_reshaped)
print(prediction)

if (prediction[0]=='R'):
  print('The object is a Rock')
else:
  print('The object is a mine')

['M']
The object is a mine
